### PR TITLE
Move modules to the `@sourcegraph/common` package into util

### PR DESCRIPTION
**Pull request by gitstart-sourcegraph**
## Decription

Move modules to the `@sourcegraph/common` package into util

## Refs
- [SourceGraph Issue](https://github.com/sourcegraph/sourcegraph/issues/29622)
- [GitStart ticket](https://app.gitstart.com/clients/sourcegraph/tickets/SG-29622)

## Implementation Detail
Nest these modules next to places where they are used under `utils` folder:
- `link-click-handler`
- `createRecord`
- `prefixSumComputer`
- `ReferenceCounter`
- `wordHelpers`

Move these modules to the `@sourcegraph/common` package into `util`:
- `browserDetection`
- `dom` — rename to `highlightNode`
- `fetchCache`
- `hashCode`
- `icon` — rename to `isEncodedImage`
- `jsonc`
- `LocalStorageSubject`
- `markdown`
- `strings.ts`
- `types.ts`
- `url.ts` — move only utilities that are not product/api specific.
- `memoizeObservable` — move to the `rxjs` folder
- `rxjs`
- `rxjs/throttleTimeWindow.ts` — it seems it can be removed.

### CODE REVIEW

**Comment by gitstart-sourcegraph** [comment]
@valerybugakov 
To confirm Implementation detail,`Nest these modules next to places where they are used under utils folder`
Entails creating new `utils` folders in directories of usage, and moving modules into them?

**Comment by valerybugakov** [comment]
> Entails creating new utils folders in directories of usage and moving modules into them?
Correct.

**Comment by gitstart-sourcegraph** [comment]
Hi @valerybugakov we noticed some issues due to importing components from `shared` package into the `common` package, causing a circular import between between @sourcegraph/shared <-> @sourcegraph/common, what do you think about best way to handle this use case

**Review by valerybugakov — CHANGES_REQUESTED** [review]

File: client/common/src/util/url.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import { tryCatch } from '@sourcegraph/common'
```
Comment (valerybugakov) [inline client/common/src/util/url.ts:1]: Let's use relative imports for modules inside of the same package.

File: client/shared/src/util/rxjs/throttleTimeWindow.ts [hunk 2]
```diff
@@ -1,127 +0,0 @@
...
- *
- * @param windowDuration Time to wait before emitting another value after emitting the last value
- * allowed by `valuesPerWindow`, measured in milliseconds or the time unit determined internally by
- * the optional `scheduler`.
- * @param valuesPerWindow The number of values to allow per time window.
- * @param The {@link SchedulerLike} to use for managing the timers that handle the throttling.
- * @returns An Observable that performs the throttle operation to limit the rate of emissions from
- * the source.
- */
-export function throttleTimeWindow<T>(
```
Comment (valerybugakov) [inline client/shared/src/util/rxjs/throttleTimeWindow.ts:32]: Nice!

**Review by valerybugakov — APPROVED** [review]
Let's resolve merge conflicts, ensure that CI is still green, and proceed with merge.
Thanks!

**Pull request merged**